src/ngraph/op/experimental/dyn_slice.cpp

```diff
@@ -0,0 +1,95 @@
+{
+    constructor_validate_and_infer_types();
+}
+
+op::DynSlice::DynSlice(const shared_ptr<Node>& arg,
+                       const shared_ptr<Node>& lower_bounds,
+                       const shared_ptr<Node>& upper_bounds)
+    : Op("DynSlice", check_single_output_args({arg, lower_bounds, upper_bounds}))
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::DynSlice::validate_and_infer_types()
+{
+    // If we only have 3 args, it means strides should have default value 1.
+    if (get_input_size() == 3)
+    {
+        // TODO: need DynConstant?
+        //m_strides = make_shared<ngraph::op::DynConstant>(element::Type_t::i64,
+        //                                                 PartialShape::dynamic(),
+        //                                                 {1});
+        throw ngraph_error("Constant strides not implemented for DynSlice");
+    }
+    auto lower_bounds_et = get_input_element_type(1);
+    auto upper_bounds_et = get_input_element_type(2);
+    auto strides_et = get_input_element_type(3);
+
+    // check data types
+    NODE_VALIDATION_CHECK(this,
+                          lower_bounds_et.compatible(element::Type_t::i64),
+                          "lower_bounds element type should be i64.");
+    NODE_VALIDATION_CHECK(this,
+                          upper_bounds_et.compatible(element::Type_t::i64),
+                          "upper_bounds element type should be i64.");
+    NODE_VALIDATION_CHECK(
+        this, strides_et.compatible(element::Type_t::i64), "strides element type should be i64.");
+
+    // check shapes
+    auto arg_rank = get_input_partial_shape(0).rank();
+    auto lower_bounds_rank = get_input_partial_shape(1).rank();
+    auto upper_bounds_rank = get_input_partial_shape(2).rank();
+    auto strides_rank = get_input_partial_shape(3).rank();
+    NODE_VALIDATION_CHECK(this,
+                          lower_bounds_rank.compatible(1),
+                          "lower_bounds should have rank 1, got ",
+                          lower_bounds_rank,
+                          ".");
+    NODE_VALIDATION_CHECK(this,
+                          upper_bounds_rank.compatible(1),
+                          "upper_bounds should have rank 1, got ",
+                          upper_bounds_rank,
+                          ".");
+    NODE_VALIDATION_CHECK(
+        this, strides_rank.compatible(1), "strides should have rank 1, got ", strides_rank, ".");
+
+    set_output_type(0, get_input_element_type(0), PartialShape::dynamic(arg_rank));
+}
+
+shared_ptr<Node> op::DynSlice::copy_with_new_args(const NodeVector& new_args) const
+{
+    check_new_args_count(this, new_args);
+    return make_shared<DynSlice>(new_args.at(0), new_args.at(1), new_args.at(2), new_args.at(3));
+}
+
+void op::DynSlice::generate_adjoints(autodiff::Adjoints& adjoints, const NodeVector& deltas)
+{
+    throw ngraph_error("generate_adjoints not implemented for DynSlice");
```

Comment:
It's likely that we will need this, so we'll want to think about the right way to implement it. (Probably a `DynReplaceSlice` will work here.)

Reply:
Sure, let's add that to the list.